app/lib/shared/deps_graph.dart

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:logging/logging.dart';
+
+import 'package:pub_dartlang_org/frontend/models.dart';
+import 'package:pub_dartlang_org/shared/task_scheduler.dart';
+
+final Logger _logger = new Logger('pub.package_graph');
+
+// To avoid having the same package names many times in memory we canonicalize them.
+class StringCanonicalization {
+  final Map<String, String> _map = <String, String>{};
+
+  String canonicalize(String value) {
+    // Fast case is when [value] is already in the map.
+    final canonicalized = _map[value];
+    if (canonicalized != null) return canonicalized;
+
+    return _map[value] = value;
+  }
+}
+
+// TODO(kustermann): We could incorporate the pubspec constraints to make the
+// sets smaller.
+class TransitiveDependencyGraph {
+  final Map<String, Set<String>> _deps = <String, Set<String>>{};
+
+  void addAll(String node, Set<String> deps) {
+    final nodes = _deps.putIfAbsent(node, _newSet);
+    final int before = nodes.length;
+    nodes.addAll(deps);
+    if (before != nodes.length) {
+      _fixpointNode(node);
+    }
+  }
+
+  void add(String node, String dep) {
+    if (_deps.putIfAbsent(node, _newSet).add(dep)) {
+      _fixpointNode(node);
+    }
+  }
+
+  Set<String> transitiveNodes(String node) => _deps.putIfAbsent(node, _newSet);
```

Comment:
The naming of the methods here doesn't make it clear in which directions it is working and what this returns (dependees vs. dependencies).